core/lib/dal/src/consensus_dal.rs

```diff
@@ -410,40 +406,31 @@ impl ConsensusDal<'_, '_> {
     /// Insertion is allowed even if it creates gaps in the L1 batch history.
     ///
     /// This method assumes that all payload validation has been carried out by the caller.
-    pub async fn insert_batch_certificate(
-        &mut self,
-        cert: &attester::BatchQC,
-    ) -> Result<(), InsertCertificateError> {
-        let l1_batch_number = cert.message.number.0 as i64;
-
-        let res = sqlx::query!(
+    /// Verification cannot be performed internally, due to circular dependency on
+    /// `zksync_l1_contract_interface`.
+    pub async fn insert_batch_certificate(&mut self, cert: &attester::BatchQC) -> DalResult<()> {
+        sqlx::query!(
             r#"
             INSERT INTO
                 l1_batches_consensus (l1_batch_number, certificate, created_at, updated_at)
             VALUES
                 ($1, $2, NOW(), NOW())
-            ON CONFLICT (l1_batch_number) DO NOTHING
             "#,
-            l1_batch_number,
+            cert.message.number.0 as i64,
             zksync_protobuf::serde::serialize(cert, serde_json::value::Serializer).unwrap(),
         )
         .instrument("insert_batch_certificate")
         .report_latency()
         .execute(self.storage)
         .await?;
-
-        if res.rows_affected().is_zero() {
-            tracing::debug!(l1_batch_number, "duplicate batch certificate");
-        }
```

Comment:
I'm not entirely comfortable removing this handling of duplicate certificate insertions. The rest of the code can still produce duplicates due to to the way re-signing works. If we merge this, that can result in an error until something else is changed. 

I might be wrong but I can also think of a convoluted scenario where a duplicate might naturally occur:
1. Say an external node collects a QC for batch N and saves it to its database. 
2. The external node is restarted and asks the main node which is the next batch number we are still voting on. 
3. Say the main node has not collected the QC for N yet (because gossip) and returns N. 
4. All other gossip partners who still have the votes for N propagate it to our external node. The QC is collected again and inserted into the database again, which returns an error now.

IMO it should either ignore it or update it, but not result in an error.

Reply:
ack, if the implementation is not compatible, then I'll leave it as it is. The silent noop behavior needs to be documented though.

Reply:
Thanks. You're right the docs should mention it. The [tests](https://github.com/matter-labs/zksync-era/blob/1c443e5ecfd000279830262a4a35cbc83a9aacec/core/lib/dal/src/consensus_dal.rs#L566-L572) only verify that there is no error raise, and Roman asked for it to be logged, but it was his or Bruno's suggestion to ignore duplicates, so I forgot to mention it as it was known to everyone, my bad.